File: src/core/python/vision_model_base.py
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
Base Model Class for All Vision Models
"""

from abc import ABC, abstractmethod
from typing import Any, Dict, Optional, Tuple
from pathlib import Path
import numpy as np


def _get_vision_root() -> Path:
    # This file lives at: <...>/cv/src/core/python/vision_model_base.py
    # Path(__file__).parents[3] == <...>/cv (project root)
    return Path(__file__).resolve().parents[3]


def _download_script_hint() -> str:
    return str(_get_vision_root() / "examples" / "<model>" / "scripts" / "download_models.sh")


def _resolve_and_check_model_path(model_path: str) -> str:
    if not model_path:
        raise FileNotFoundError("Model path is empty")
    if model_path.startswith("http://") or model_path.startswith("https://"):
        raise FileNotFoundError(
            f"Model URL is not supported: {model_path}. Please download the model first, e.g. run: "
            f"{_download_script_hint()}"
        )
    p = Path(model_path).expanduser()
    if not p.is_absolute():
        p = (_get_vision_root() / p).resolve()
    if not p.exists():
        raise FileNotFoundError(
            f"Model file not found: {p}. Please download the model first, e.g. run: {_download_script_hint()}"
        )
    return str(p)


class BaseModel(ABC):
    """
    Abstract base class for all vision models.

    All vision models should inherit from this class and implement the required methods.
    """

    def __init__(self, model_path: str, lazy_load: bool = False, **kwargs):
        """
        Initialize the model.

        Args:
            model_path: Path to the model file (e.g., ONNX, TorchScript)
            lazy_load: If True, delay model loading until first inference call
            **kwargs: Additional model-specific parameters
        """
        self.model_path = _resolve_and_check_model_path(model_path)
        self.session = None
        self.input_shape = None
        self._model_loaded = False
        self._lazy_load = lazy_load
        self._load_kwargs = kwargs

        if not lazy_load:
            self._load_model(**kwargs)
            self._model_loaded = True

    @abstractmethod
    def _load_model(self, **kwargs):
        """
        Load the model from file.

        This method should:
        1. Initialize the inference session
        2. Get input/output information
        3. Store model metadata
        """
        pass

    @abstractmethod
    def preprocess(self, image: np.ndarray) -> Any:
        """
        Preprocess the input image for inference.

        Args:
            image: Input image in BGR format (OpenCV default)

        Returns:
            Preprocessed input ready for model inference
        """
        pass

    def _ensure_model_loaded(self):
        """Ensure model is loaded (for lazy loading)."""
        if not self._model_loaded:
            self._load_model(**self._load_kwargs)
            self._model_loaded = True

    @abstractmethod
    def infer(self, image: np.ndarray) -> Any:
        """
        Run inference on the input image.

        Args:
            image: Input image in BGR format

        Returns:
            Model output (format depends on specific model type)
        """
        pass

    def postprocess(self, outputs: Any, **kwargs) -> Any:
        """
        Postprocess the model outputs.

        Args:
            outputs: Raw model outputs
            **kwargs: Additional parameters for postprocessing

        Returns:
            Processed results in a usable format
        """
        pass

    def get_input_shape(self) -> Optional[Tuple[int, ...]]:
        """
        Get the input shape of the model.

        Returns:
            Input shape tuple (C, H, W) or None if not available
        """
        return self.input_shape

    def get_model_info(self) -> Dict[str, Any]:
        """
        Get model information.

        Returns:
            Dictionary containing model metadata
        """
        return {
            'model_path': self.model_path,
            'input_shape': self.input_shape,
        }

File: src/core/cpp/vision_model_base.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "vision_model_base.h"

#include <chrono>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <memory>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#include <filesystem>  // NOLINT(build/c++17)

namespace vision_core {

namespace fs = std::filesystem;

static bool is_url(const std::string& path) {
    return path.rfind("http://", 0) == 0 || path.rfind("https://", 0) == 0;
}

static fs::path get_vision_root_dir() {
#ifdef CV_MODELZOO_VISION_ROOT
    return fs::path(CV_MODELZOO_VISION_ROOT);
#else
    return fs::current_path();
#endif
}
static std::string download_script_hint() {
    const fs::path hint =
        get_vision_root_dir() / "examples" / "<model>" / "scripts" / "download_models.sh";
    return hint.lexically_normal().string();
}

static std::string expand_tilde(const std::string& path) {
    if (path.empty() || path[0] != '~') return path;
    const char* home = std::getenv("HOME");
    if (!home || home[0] == '\0') return path;
    if (path.size() == 1 || path[1] == '/')
        return std::string(home) + path.substr(1);
    return path;
}

static std::string resolve_and_check_model_path(const std::string& raw_path) {
    if (raw_path.empty()) {
        throw std::runtime_error("Model path is empty");
    }
    if (is_url(raw_path)) {
        std::ostringstream oss;
        oss << "Model URL is not supported: " << raw_path
            << ". Please download the model first, e.g. run: " << download_script_hint();
        throw std::runtime_error(oss.str());
    }

    std::string expanded = expand_tilde(raw_path);
    fs::path p(expanded);
    p = p.lexically_normal();

    fs::path candidate;
    if (p.is_absolute()) {
        candidate = p;
        if (fs::exists(candidate)) {
            return candidate.string();
        }
    } else {
        candidate = fs::current_path() / p;
        candidate = candidate.lexically_normal();
        if (fs::exists(candidate)) {
            return candidate.string();
        }

        candidate = get_vision_root_dir() / p;
        candidate = candidate.lexically_normal();
        if (fs::exists(candidate)) {
            return candidate.string();
        }
    }

    {
        std::ostringstream oss;
        oss << "Model file not found: " << p.string()
            << ". Please download the model first, e.g. run: " << download_script_hint();
        throw std::runtime_error(oss.str());
    }
}

BaseModel::BaseModel(const std::string& model_path, bool lazy_load)
    : model_path_(model_path), model_loaded_(false), lazy_load_(lazy_load) {
    env_ = std::make_unique<Ort::Env>(ORT_LOGGING_LEVEL_WARNING, "VisionModel");

    if (!lazy_load) {
        (void)0;  // Subclass may load in constructor
    }
}

BaseModel::~BaseModel() {
    release();
}

void BaseModel::warmup() {
    ensure_model_loaded();
}

void BaseModel::release() {
    session_.reset();
    input_shape_.clear();
    input_node_names_.clear();
    output_node_names_.clear();
    input_names_.clear();
    output_names_.clear();
    output_num_ = 0;
    model_loaded_ = false;
}

std::vector<ModelCapability> BaseModel::get_capabilities() const {
    return {ModelCapability::kImageInput};
}

bool BaseModel::supports_capability(ModelCapability capability) const {
    const std::vector<ModelCapability> capabilities = get_capabilities();
    for (const auto& cap : capabilities) {
        if (cap == capability) {
            return true;
        }
    }
    return false;
}

std::vector<int64_t> BaseModel::get_input_shape() const {
    return input_shape_;
}

std::string BaseModel::get_model_info() const {
    std::string info = "Model Path: " + model_path_ + "\n";
    info += "Input Shape: ";
    for (size_t i = 0; i < input_shape_.size(); ++i) {
        info += std::to_string(input_shape_[i]);
        if (i < input_shape_.size() - 1) {
            info += ", ";
        }
    }
    return info;
}

RuntimeProfile BaseModel::get_runtime_profile() const {
    return runtime_profile_;
}

void BaseModel::reset_runtime_profile() {
    runtime_profile_ = RuntimeProfile{};
}

void BaseModel::set_runtime_preprocess_ms(double ms) {
    runtime_profile_.preprocess_ms = ms;
}

void BaseModel::set_runtime_model_infer_ms(double ms) {
    runtime_profile_.model_infer_ms = ms;
}

void BaseModel::set_runtime_postprocess_ms(double ms) {
    runtime_profile_.postprocess_ms = ms;
}

void BaseModel::set_runtime_detect_ms(double ms) {
    runtime_profile_.detect_ms = ms;
}

void BaseModel::set_runtime_track_ms(double ms) {
    runtime_profile_.track_ms = ms;
}

void BaseModel::set_runtime_total_ms(double ms) {
    runtime_profile_.total_ms = ms;
}

void BaseModel::ensure_model_loaded() {
    if (!model_loaded_) {
        load_model();
        model_loaded_ = true;
    }
}

void BaseModel::init_session(int num_threads, const std::string& provider) {
    model_path_ = resolve_and_check_model_path(model_path_);

    Ort::SessionOptions session_options;
    session_options.SetIntraOpNumThreads(num_threads);
    session_options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);

    if (provider == "SpaceMITExecutionProvider") {
        Ort::Status status = Ort::SessionOptionsSpaceMITEnvInit(session_options);
        if (status.IsOK()) {
            std::cout << "SpaceMIT EP initialized successfully!" << std::endl;
        } else {
            std::cerr << "SpaceMIT EP init failed: " << status.GetErrorMessage() << std::endl;
        }
    }

    session_ = std::make_unique<Ort::Session>(*env_, model_path_.c_str(), session_options);

    size_t num_inputs = session_->GetInputCount();
    input_node_names_.resize(num_inputs);
    input_names_.resize(num_inputs, "");

    for (size_t i = 0; i < num_inputs; ++i) {
        auto input_name = session_->GetInputNameAllocated(i, allocator_);
        input_names_[i] = input_name.get();
        input_node_names_[i] = input_names_[i].c_str();
    }

    Ort::TypeInfo input_type_info = session_->GetInputTypeInfo(0);
    auto input_tensor_info = input_type_info.GetTensorTypeAndShapeInfo();
    input_shape_ = input_tensor_info.GetShape();

    size_t num_outputs = session_->GetOutputCount();
    output_node_names_.resize(num_outputs);
    output_names_.resize(num_outputs, "");
    output_num_ = num_outputs;

    for (size_t i = 0; i < num_outputs; ++i) {
        auto output_name = session_->GetOutputNameAllocated(i, allocator_);
        output_names_[i] = output_name.get();
        output_node_names_[i] = output_names_[i].c_str();
    }

    memory_info_ = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
}

std::vector<Ort::Value> BaseModel::run_session(const cv::Mat& input_blob) {
    ensure_model_loaded();

    if (input_blob.empty()) {
        throw std::runtime_error("Input blob is empty");
    }

    std::vector<int64_t> tensor_shape = input_shape_;

    if (!tensor_shape.empty() && tensor_shape[0] <= 0) {
        tensor_shape[0] = 1;
    }

    Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
        memory_info_,
        const_cast<float*>(input_blob.ptr<float>()),
        input_blob.total(),
        tensor_shape.data(),
        tensor_shape.size());

    const auto t0 = std::chrono::steady_clock::now();
    std::vector<Ort::Value> outputs = session_->Run(
        Ort::RunOptions{nullptr},
        input_node_names_.data(),
        &input_tensor,
        1,
        output_node_names_.data(),
        output_node_names_.size());
    const auto t1 = std::chrono::steady_clock::now();
    runtime_profile_.model_infer_ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    return outputs;
}

}  // namespace vision_core
